### PR TITLE
feat: core.331 support.gcs.artifact.repository

### PR DESCRIPTION
**Pull request by aleksandrmelnikov**
**What this PR does**:
Updated code to support Google cloud storage if available.

**Which issue(s) this PR fixes**:

Fixes onepanelio/core#331

**Special notes for your reviewer**:

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -30,10 +30,14 @@ require (
...
 	github.com/stretchr/testify v1.4.0
 	github.com/tmc/grpc-websocket-proxy v0.0.0-20200122045848-3419fae592fc
+	golang.org/x/net v0.0.0-20200301022130-244492dfa37a
+	golang.org/x/oauth2 v0.0.0-20200107190931-bf48bf16ab8d
+	google.golang.org/api v0.20.0
 	google.golang.org/genproto v0.0.0-20200430143042-b979b6f78d84
 	google.golang.org/grpc v1.28.0
 	google.golang.org/protobuf v1.22.0
 	gopkg.in/yaml.v2 v2.2.8
+	gopkg.in/yaml.v3 v3.0.0-20200615113413-eeeca48fe776
```
Comment (rushtehrani) [inline go.mod:40]: Why do we need this too? Now we have are using 4 different libraries. Can you use `"sigs.k8s.io/yaml"` instead? I'd like to eventually refactor everything to use that.

File: pkg/workflow_execution.go [hunk 2]
```diff
@@ -889,30 +933,61 @@ func (c *Client) GetWorkflowExecutionMetrics(namespace, uid, podName string) (me
...
+					"UID":       uid,
+					"PodName":   podName,
+					"Error":     err.Error(),
+				}).Error("Metrics do not exist.")
+				return nil, util.NewUserError(codes.NotFound, "Metrics do not exist.")
+			}
+		}
+	case config.ArtifactRepository.GCS != nil:
+		{
+			ctx := context.Background()
```
Comment (rushtehrani) [inline pkg/workflow_execution.go:966]: [These lines](https://github.com/onepanelio/core/pull/400/files#diff-2df54c8de22b7ec30c64350c5cbfd591R966-R987) seem to be repeated a bunch of times, can this be put into one method?

File: go.mod (hunk 1 above)
Comment (aleksandrmelnikov) [reply go.mod:40]: Going to create a separate ticket to investigate this.

- We use yaml.v3 to support reading (and writing) YAML with comments, inline comments, and other metadata.
- We use a pkg/config_types.go in CLI to marshal / unmarshal `params.yaml`
- Same package is used in core to dynamically generate the s3 access and secret keys, for Workflows

I think, ideally, we should use yaml.v3

Comment (aleksandrmelnikov) [reply go.mod:40]: https://github.com/onepanelio/core/issues/409

File: pkg/workflow_execution.go (hunk 2 above)
Comment (aleksandrmelnikov) [reply pkg/workflow_execution.go]: Refactored

File: pkg/config_types.go [hunk 3]
```diff
@@ -137,15 +138,115 @@ func (s SystemConfig) DatabaseConnection() (driverName, dataSourceName string) {
...
 	Bucket          string
 	Endpoint        string
 	Insecure        bool
 	Region          string
 	AccessKeySecret corev1.SecretKeySelector
 	SecretKeySecret corev1.SecretKeySelector
-	AccessKey       string
-	Secretkey       string
+	AccessKey       string `yaml:"accessKey"`
+	Secretkey       string `yaml:"secretKey"`
```
Comment (rushtehrani) [inline pkg/config_types.go:149]: should be `SecretKey`

Comment (rushtehrani) [reply pkg/config_types.go:149]: Do we need this if we have `ArtifactRepositoryS3Provider`? Can't we just move `AccessKey` and `SecretKey` fields to `ArtifactRepositoryS3Provider` and use that across the board?

File: pkg/config_types.go [hunk 4]
```diff
@@ -162,8 +263,23 @@ func (a *ArtifactRepositoryS3Config) FormatKey(namespace, workflowName, podName
...
+	keyFormat := g.KeyFormat
+
+	keyFormat = strings.Replace(keyFormat, "{{workflow.namespace}}", namespace, -1)
+	keyFormat = strings.Replace(keyFormat, "{{workflow.name}}", workflowName, -1)
+	keyFormat = strings.Replace(keyFormat, "{{pod.name}}", podName, -1)
+
+	return keyFormat
+}
+
 type ArtifactRepositoryConfig struct {
```
Comment (rushtehrani) [inline pkg/config_types.go:280]: Can't we just use `ArtifactRepositoryProviderConfig` and make the fields pointers there?

File: pkg/config_types.go (hunk 3 above)
Comment (aleksandrmelnikov) [reply pkg/config_types.go:149]: > should be `SecretKey`

Sorry, do you mean the `yaml: "secretKey"`?

Comment (rushtehrani) [reply pkg/config_types.go:149]: @aleksandrmelnikov the `K` in the actual field name is lower case now, should be uppercase.

Comment (aleksandrmelnikov) [reply pkg/config_types.go]: @rushtehrani I was able to move the attributes from S3Config to S3Provider.

Did you to remove the S3Config struct when you said "use across the board"?

File: pkg/config_types.go (hunk 4 above)
Comment (aleksandrmelnikov) [reply pkg/config_types.go]: Yes, this makes sense.

File: pkg/config_types.go (hunk 3 above)
Comment (aleksandrmelnikov) [reply pkg/config_types.go]: Made changes.

**Pull request merged**